AEPMessaging/Sources/Messaging.swift

```diff
@@ -633,59 +626,72 @@ public class Messaging: NSObject, Extension {
+        if surfaceRulesBySchemaType[.contentCard] != nil {
+            let collectedContentCardRules = collectRules(from: contentCardRulesBySurface)
+            contentCardRulesEngine.launchRulesEngine.replaceRules(with: collectedContentCardRules)
 
-                // pre-fetch the assets for this message if there are any defined
-                rulesEngine.cacheRemoteAssetsFor(allInAppRules)
+            // Seed content cards qualification
+            let seedEvent = Event(name: "Seed content cards", type: EventType.messaging, source: EventSource.requestContent, data: nil)
+            let qualified = getPropositionsFromContentCardRulesEngine(seedEvent)
+            for (surface, propositions) in qualified {
+                addOrReplaceContentCards(propositions, forSurface: surface)
+            }
+        }
 
-                // update rules in in-app engine
-                rulesEngine.launchRulesEngine.replaceRules(with: allInAppRules)
+        // In-app + event history rules engine
+        if surfaceRulesBySchemaType[.inapp] != nil || surfaceRulesBySchemaType[.eventHistoryOperation] != nil {
+            let collectedInAppRules = collectRules(from: inAppRulesBySurface)
 
-            case .feed, .contentCard:
-                Log.trace(label: MessagingConstants.LOG_TAG, "Updating content card definitions for surfaces \(newRules.compactMap { $0.key.uri }).")
+            // Prefetch assets only if we actually received in-app rules this time
+            if surfaceRulesBySchemaType[.inapp] != nil {
+                rulesEngine.cacheRemoteAssetsFor(collectedInAppRules)
+            }
 
-                // replace rules for each content card surface we got back
-                contentCardRulesBySurface.merge(newRules) { _, new in new }
+            // Combine in-app + event history rules
+            var combined = collectedInAppRules
+            combined.append(contentsOf: collectRules(from: eventHistoryRulesBySurface))
 
-                // remove any surfaces that were requested but had no in-app content returned
-                for surface in surfacesToRemove {
-                    contentCardRulesBySurface.removeValue(forKey: surface)
-                }
+            rulesEngine.launchRulesEngine.replaceRules(with: combined)
+        }
+    }
 
-                // update rules in content card rules engine
-                contentCardRulesEngine.launchRulesEngine.replaceRules(with: contentCardRulesBySurface.flatMap { $0.value })
+    private func processRulesForSchemaType(_ surfaceRulesBySchemaType: [SchemaType: [Surface: [LaunchRule]]], requestedSurfaces: [Surface], schemaType: SchemaType, rulesBySurface: inout [Surface: [LaunchRule]]) {
+        if let newRules = surfaceRulesBySchemaType[schemaType] {
+            let newSurfaces = Array(newRules.keys)
+            Log.trace(label: MessagingConstants.LOG_TAG, "Updating definitions for surfaces \(newSurfaces.map { $0.uri }) with schema type \(schemaType).")
```

Comment:
i guess technically we are updating definitions for all requested surfaces, not just those that came back in the response